src/main/java/com/aws/greengrass/device/session/SessionManager.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.device.session;
+
+import com.aws.greengrass.device.exception.AuthenticationException;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.util.Pair;
+import lombok.AccessLevel;
+import lombok.Getter;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Singleton class for managing AuthN and AuthZ session.
+ */
+public class SessionManager {
+    private static final Logger logger = LogManager.getLogger(SessionManager.class);
+    public static final String SESSION_ID = "SessionId";
+
+    // There are two types of Session IDs defined here. Internal session IDs
+    // are derived by hashing device credentials. This is done so that we can
+    // avoid creating multiple sessions for the same device. Only a single session
+    // for a given set of device credentials is allowed at a given time.
+    // However, Session IDs should not be predictable. Here, we create a separate
+    // external Session ID which can be returned to the IPC client. The external
+    // Session ID is stored along with the Session itself so that we can clean up
+    // duplicate external Session IDs
+    @Getter(AccessLevel.PACKAGE)
+    private final Map<String, Pair<String,Session>> sessionMap = new ConcurrentHashMap<>();
+    private final Map<String, String> externalToInternalSessionMap = new ConcurrentHashMap<>();
+
+    /**
+     * find session by id.
+     *
+     * @param sessionId session identifier
+     * @return session or null
+     */
+    public Session findSession(String sessionId) {
+        String internalSessionId = externalToInternalSessionMap.get(sessionId);
```

Comment:
why do we need indirection of internal and external session IDs?

Reply:
I could be convinced that it's not needed. But I'm hesitant to return a "predictable" session ID to clients - similar to guidance given for pagination tokens. Worth a quick discussion with the security team, I suppose

Reply:
I'm ok with it, but if removing the indirection will simplify anything then I'd push for it. Pagination token needs to be kept more obscure AFAIK because it can directly include the actual value of the last scanned key, so you're leaking info. Just having a session ID which is a hash of the inputs should be opaque enough.

Reply:
You're right, the reasons for opaque pagination tokens don't quite apply here. I think the complexity can be contained within this class, so I'm not too too concerned with it. But we should definitely think harder if it starts leaking out.

Other reasons for returning unique IDs:
1. Improved debuggability/tracability - I can see not being able to differentiate between two sessions created with the same credentials being annoying
2. Less chance of a client side bug (using wrong auth token to authorize an action)

We should at least make sure the format is the same, however. I'll make a change to fix that so we have the option to easily change this behavior in the future without risk of breaking anything.

Reply:
I'm fine with it being uuids, no concern there.

Reply:
I removed this logic for now. The code path that Moquette uses currently creates sessions based on certificates, and we can't de-dup based on this alone (since multiple clients could use the same cert)